### PR TITLE
Drop deprecated auth helpers from Connection and refactor auth middleware

### LINKED ISSUE #1302: [2.0] Deprecations to drop
This PR is a "quest issue" about removing deprecated functionality in the next major version.

## Description

We have some already-described deprecations, and can now remove them.

If you find more deprecated functionality, please add it to this list, which is meant to guide the 2.0 release.

- [x] FilePart: Drop the "consistency between UploadIO" method  https://github.com/lostisland/faraday/blob/21ac595919b0372b41386ae00640e16f73a20d31/lib/faraday/file_part.rb#L57
- [x] #1303: Drop the `method` alias for `http_method`.
- [ ] #1308: Authorization: drop the helpers available in `Faraday::Connection`

### PR DESCRIPTION
**Pull request by iMacTia**
## Description
* Drop deprecated auth helpers from Connection
* Merge Authentication middleware.
* Add possibility of passing a proc.
* Update documentation.

Fixes #1302 as this is the last deprecation in the 1.x code 🙌 

## Todos
List any remaining work that needs to be done, i.e:
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: spec/faraday/request/authorization_spec.rb [hunk 1]
```diff
@@ -14,18 +14,18 @@
 
   shared_examples 'does not interfere with existing authentication' do
     context 'and request already has an authentication header' do
-      let(:response) { conn.get('/auth-echo', nil, authorization: 'Token token="bar"') }
+      let(:response) { conn.get('/auth-echo', nil, authorization: 'OAuth oauth_token') }
```
Comment (olleolleolle) [inline spec/faraday/request/authorization_spec.rb:17]: Nice improvement!

**Review by olleolleolle — APPROVED** [review]
Good removal! 🥳 

This brings clarity to a place where the story was not so straight-forward to tell before!

**Comment by etiennebarrie** [comment]
This broke passing a URL with embedded basic auth in the user info part of the URI, because basic_auth is still used in `url_prefix=` https://github.com/lostisland/faraday/blob/0f9626c48d0daa24888cb4e5e7962c106a48d97f/lib/faraday/connection.rb#L364-L367

```
~/src/github.com/lostisland/faraday (1.x) $ ruby -Ilib -rfaraday -rfaraday/net_http -rjson -e "puts JSON.load(Faraday.new(url: 'https://user:password@httpbin.org/headers').get.body)['headers']['Authorization']"
WARNING: `Faraday::Connection#basic_auth` is deprecated; it will be removed in version 2.0.
While initializing your connection, use `#request(:basic_auth, ...)` instead.
See https://lostisland.github.io/faraday/middleware/authentication for more usage info.
Basic dXNlcjpwYXNzd29yZA==
```

```
~/src/github.com/lostisland/faraday (main) $ ruby -Ilib -rfaraday -rfaraday/net_http -rjson -e "puts JSON.load(Faraday.new(url: 'https://user:password@httpbin.org/headers').get.body)['headers']['Authorization']"
lib/faraday/connection.rb:365:in `block in url_prefix=': undefined method `basic_auth' for #<Faraday::Connection:0x00007fed250abaa8 @parallel_manager=nil, @headers={}, @params={}, @options=#<Faraday::RequestOptions (empty)>, @ssl=#<Faraday::SSLOptions (empty)>, @default_parallel_manager=nil, @manual_proxy=nil, @builder=#<Faraday::RackBuilder:0x00007fed25092210 @adapter=Faraday::Adapter::NetHttp, @handlers=[Faraday::Request::UrlEncoded]>, @url_prefix=#<URI::HTTPS https://user:password@httpbin.org/headers>> (NoMethodError)
	from lib/faraday/connection.rb:506:in `with_uri_credentials'
	from lib/faraday/connection.rb:364:in `url_prefix='
	from lib/faraday/connection.rb:84:in `initialize'
	from lib/faraday.rb:96:in `new'
	from lib/faraday.rb:96:in `new'
	from -e:1:in `<main>'
```

While I'm not against deprecating `basic_auth` and `authorization` on Connection, I think being able to have basic auth embedded in the URL is really useful (it can be configured in secrets for example).

**Comment by iMacTia** [comment]
Thanks for reporting this issue @etiennebarrie, I'm surprised we didn't have tests supporting this use-case.
I've opened an issue (#1319) from your comment and will discuss with the team on the course of action, please feel free to join the discussion there

**Pull request merged**